explorer: add analyze option

Ana06 opened the pull request:
I would like to load capa explorer with an script and that it runs the
analysis without needing extra clicks. Introduce an analyze option for
this.

Loading capa explorer from the UI or with Alt+F5 behaves as before. The
following command as well:
```
ida_loader.load_and_run_plugin("capa_explorer", 0)
```
But the following command automatically runs the analysis without extra
clicks:
```
ida_loader.load_and_run_plugin("capa_explorer", 1)
```

Example of where I am using this:
https://github.com/Ana06/idapython/blob/master/idapythonrc.py#L22

Closes https://github.com/fireeye/capa/issues/546

### Documentation

- [x] I have updated the [CHANGELOG.md](/CHANGELOG.md), this is required for:
  - Bug fixes (non-breaking change which fixes an issue)
  - New features (non-breaking change which adds functionality)
  - Breaking changes (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
  - [x] I have made the corresponding changes to the documentation

### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] No new tests needed

williballenthin commented on capa/ida/plugin/form.py:
```diff
@@ -227,7 +227,7 @@ def get_values(self):
     """form element for plugin interface"""
 
-    def __init__(self, name):
+    def __init__(self, name, option=0):
```
i'd recommend:
  - a boolean for this, rather than integer
  - a human readable name

Ana06 replied:
`run` receives a number: https://www.hex-rays.com/blog/scriptable-plugins

I can change the name by `analyze` or something else. But I was thinking we may want to add other options. For example: https://github.com/fireeye/capa/issues/547

🤔

williballenthin replied:
ah, ok, thanks for the explanation. maybe a tiny comment to this effect, then?

if we're going to support multiple options, but only a single arg, maybe we could use a bitmask here?

mr-tz replied:
yes and use constants a enum?

Ana06 replied:
Yes! bitmask + enum is what I was thinking too when/if there are more options. Should we use it for one as well?

I am not sure this is the best alternative (it is probably the easiest/fastest). Ideally the explorer would run the analysis for both views when opening the explorer always. Even for https://github.com/fireeye/capa/issues/547, I would prefer `select all` to be the default. If we disagree on the defaults, that would be a reason to have options.

Ana06 replied:
Yes! bitmask + enum is what I was thinking too when/if there are more options. Should we use it for one as well?

mr-tz replied:
I think so

mr-tz commented:
Analyzing on start could also be configured via the plugin settings? I'd enable it.

Ana06 commented:
@mr-tz 
> Analyzing on start could also be configured via the plugin settings? I'd enable it.

Yes, if it would be done for both views. If only the `Program Analysis` analysis is done and I want to use the `Rule Generator` it is not very convenient. If it is done for both views then IMO it can just be the default and not need a configuration option (we may not even need an analyze button). It is needed to analyze before using the explorer, so it should be IMO done without user interaction.

See https://github.com/fireeye/capa/issues/546

mike-hunhoff commented:
> @mr-tz
> 
> > Analyzing on start could also be configured via the plugin settings? I'd enable it.
> 
> Yes, if it would be done for both views. If only the `Program Analysis` analysis is done and I want to use the `Rule Generator` it is not very convenient. If it is done for both views then IMO it can just be the default and not need a configuration option (we may not even need an analyze button). It is needed to analyze before using the explorer, so it should be IMO done without user interaction.
> 
> See #546

I just added some comments to #546.

We could change capa explorer to automatically run program analysis when opened. However, I do not think we can remove  the `Analyze` button altogether. We want the user to be able to run analysis more than once without needing to close and reopen the plugin. The rule generator also needs to run analysis per function and the `Analyze` button allows users to have control over which function is being analyzed and when.

Ana06 commented:
I had deleted my comment after reading @mike-hunhoff comment (but @mike-hunhoff was too fast!). I have changed my opinion. I think having options (as in this PR) is the best approach. A default for the explorer is a bad idea because if you close the explorer and open it only for writing a rule then it runs the analysis again.

mr-tz commented:
are we planning to move forward with this?

Ana06 commented:
@mr-tz 

> are we planning to move forward with this?

I have just update it to use bitmask + enum for the option. Is there something else to discuss address?

mr-tz reviewed (approved):
looks good to me

williballenthin commented on CHANGELOG.md:
```suggestion
- explorer: add argument to control whether to automatically analyze when running capa explorer #548 @Ana06
```

williballenthin commented on capa/ida/plugin/form.py:
```diff
@@ -275,6 +282,9 @@ def __init__(self, name):
 
         self.Show()
 
+        if (option & Options.ANALYZE) == Options.ANALYZE:
```
the argument here is being used like a bitflag, but everywhere else is described as a number. i guess thats not technically wrong, but we'd need to update more places as soon as we add a second bit.

Ana06 replied:
we could also use it as a number here until we add another option. Maybe we never add a second option and this is a bit confusing with one option. That's what I meant [before](https://github.com/fireeye/capa/pull/548#discussion_r625910367).

williballenthin replied:
yes, and suggested changes above update the documentation to describe this as a bitflag.

williballenthin commented on capa/ida/plugin/form.py:
```diff
@@ -41,6 +41,13 @@
+from enum import Enum
+
+
+class Options(Enum):
```
im just learning about this myself, but i think the appropriate class here is `Flag`:
https://docs.python.org/3/library/enum.html#flag

Ana06 replied:
Actually `(option & Options.ANALYZE) == Options.ANALYZE` (the current code) doesn't work:
```
TypeError: unsupported operand type(s) for &: 'int' and 'Options'
```

I think we should use `IntFlag`. With `class Options(Flag)`, we need to cast `options` as `&` is not defined for `int` and `Flag` and we receive and `int`: `(Options(option) & Options.ANALYZE) == Options.ANALYZE`. With `class Options(IntFlag)` the casting is not needed.

We could also use `IntEnum`. I think the difference between `IntFlag` and `IntEnum` is that it allow us to define options from other options `(VERBOSE_ANALYZE = VERBOSE | ANALYZE)`. I guess it doesn't hurt to use `IntFlag` in case we want to use it as some point.

williballenthin replied:
ok, thats fine - whatever works for our situation. i was just googling around for supported solutions to bitflags in python

williballenthin commented on CHANGELOG.md:
```suggestion
  `ida_loader.load_and_run_plugin("capa_explorer", Options.ANALYZE)`
```

Ana06 replied:
users would need to define `Options.ANALYZE` in their scripts/code and I think this is suppose to be an `int`. 😕

williballenthin replied:
ok or we can export it somehow? maybe put it in `capa/ida/__init__.py`?

Ana06 replied:
Plugins can also be configured in the `plugins.cfg` file. There it needs to be a number.
Ref: https://hex-rays.com/blog/scriptable-plugins/

williballenthin reviewed (approved):
see inline recommendations to clarify wording around bitflag vs number

mr-tz commented:
let's get this merged

Ana06 commented:
I think I have addressed all Willi's comments, but I haven't tested it, as there is an unrelated bug in the explorer. 😶

mr-tz commented:
blocked by https://github.com/fireeye/capa/issues/653

mr-tz commented:
Can you test this now and then let's merge. Sorry for the changelog conflict 😐

Ana06 commented:
Tested! It works as expected 😉

mr-tz commented:
Awesome! Please merge at your convenience 😄
